tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Grpc/GrpcLegacy/Client/AsyncCallHandleUnaryResponseInstrumentation.cs

```diff
@@ -0,0 +1,102 @@
+﻿// <copyright file="AsyncCallHandleUnaryResponseInstrumentation.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.ComponentModel;
+using Datadog.Trace.ClrProfiler.AutoInstrumentation.Grpc.GrpcLegacy.Client.DuckTypes;
+using Datadog.Trace.ClrProfiler.CallTarget;
+using Datadog.Trace.Configuration;
+using Datadog.Trace.DuckTyping;
+using Datadog.Trace.ExtensionMethods;
+using Datadog.Trace.Tagging;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.Grpc.GrpcLegacy.Client
+{
+    /// <summary>
+    /// Grpc.Core.Internal calltarget instrumentation
+    /// </summary>
+    [InstrumentMethod(
+        AssemblyName = "Grpc.Core",
+        TypeName = "Grpc.Core.Internal.AsyncCall`2",
+        MethodName = "HandleUnaryResponse",
+        ReturnTypeName = ClrNames.Void,
+        ParameterTypeNames = new[] { ClrNames.Bool, "Grpc.Core.Internal.ClientSideStatus", "Grpc.Core.Internal.IBufferReader", "Grpc.Core.Metadata" },
+        MinimumVersion = "2.0.0",
+        MaximumVersion = "2.*.*",
+        IntegrationName = nameof(Grpc))]
+    [Browsable(false)]
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public static class AsyncCallHandleUnaryResponseInstrumentation
+    {
+        internal static CallTargetState OnMethodBegin<TTarget, TStatus, TBufferReader, TMetadata>(
+            TTarget instance,
+            bool success,
+            in TStatus clientSideStatus,
+            in TBufferReader receivedMessageReader,
+            in TMetadata responseHeaders)
+        {
+            // receivedStatus can change in the middle of this method, so we need to wait for the method to finish,
+            // to grab the final status.
+            var tracer = Tracer.Instance;
+            if (GrpcCoreApiVersionHelper.IsSupported && tracer.Settings.IsIntegrationEnabled(IntegrationId.Grpc))
+            {
+                // using CreateFrom to avoid boxing ClientSideStatus struct
+                var receivedStatus = DuckType.CreateCache<ClientSideStatusWithMetadataStruct>.CreateFrom(clientSideStatus);
+                var status = receivedStatus.Status;
+                var asyncCall = instance.DuckCast<AsyncCallStruct>();
+                var scope = GrpcLegacyClientCommon.CreateClientSpan(tracer, in asyncCall.Details, in status);
+                if (scope?.Span is { } span)
+                {
+                    if (receivedStatus.Trailers is { Count: > 0 })
+                    {
+                        span.SetHeaderTags(new MetadataHeadersCollection(receivedStatus.Trailers), tracer.Settings.GrpcTags, defaultTagPrefix: GrpcCommon.ResponseMetadataTagPrefix);
+                    }
+                    else if (responseHeaders is not null)
```

Comment:
How come we only apply one set of headers instead of both?

Reply:
As far as I can tell, only one of these will ever have values. In all cases I've found, the first `if` that checks `receivedStatus.Trailers` will be the one with values. By using `else if`, when this _does_ have values, we can skip the extra allocation that's needed to access `responseHeaders`. 

This is obviously a micro-optimisation, so happy to remove the `else` if consensus is that it's safer/more obvious 🙂